astropy/units/distribution.py

```diff
@@ -0,0 +1,427 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+"""
+Distribution class and associated machinery.
+"""
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import numpy as np
+
+from . import Quantity, UnitsError
+from .. import visualization
+
+__all__ = ['Distribution', 'NormalDistribution', 'PoissonDistribution',
+           'UniformDistribution']
+
+
+class Distribution:
+    """
+    A unitful value with associated uncertainty distribution.  While subclasses
+    may have exact analytic forms, this class uses samples to represent the
+    distribution.
+
+    Parameters
+    ----------
+    distr : array-like
+        The distribution, with sampling along the *leading* axis.
+    distr_center : `None` or array-like
+        The "center" of this distribution.  It must have the same shape as
+        ``distr``, aside from the (initial) sample dimension. If `None`, the
+        *median* of ``distr``  along the sampling axis will be used.
+    unit : astropy unit
```

Comment:
And remove the `unit` argument as well - people should just pass in a `Quantity` if they want units - as is, the code allows a non-unit array, and there doesn't seem anything wrong with that.

(Reminded of this because of the comments in `__new__` below.)